src/python/pants/engine/intrinsics.py

```diff
@@ -120,13 +121,39 @@ async def run_id() -> RunId:
     return await native_engine.run_id()
 
 
+# NB: Call one of the helpers below, instead of calling this rule directly,
+#  to ensure correct application of restartable logic.
 @_uncacheable_rule
-async def interactive_process(
+async def _interactive_process(
     process: InteractiveProcess, process_execution_environment: ProcessExecutionEnvironment
 ) -> InteractiveProcessResult:
     return await native_engine.interactive_process(process, process_execution_environment)
 
 
+async def run_interactive_process(iproc: InteractiveProcess) -> InteractiveProcessResult:
+    # NB: We must call mark_nonrestartable() in this helper, rather than in a nested @rule call,
+    #  so that the Task for the @rule that calls this helper is the one marked as non-restartable.
+    if not iproc.restartable:
+        mark_nonrestartable()
```

Comment:
I feel like the interactions here are worth more discussion, because this feels super-subtle!

Potentially even a reference in the `intrinsics.rs` code/on `mark_nonrestartable`

Reply:
It is super subtle, but not really more than it was before, except that now the subtlety is in Python instead of Rust.

Reply:
Re a comment on `mark_nonrestartable()` (which is now gone since we can use the existing `task_side_effected()`) - I'm not sure that's the right place to introduce this subtlety. `task_side_effected()` does exactly what it says on the tin. The subtlety is that we have to ensure we call it on the right task... Which is an issue relevant here, but not there. And it is also only an issue when this specific intrinsic is called from Python. So this seems like the right place for the comment. But if you have suggestions for further elaborating in this comment - glad to augment!